server/schedule/operator/operator.go

```diff
@@ -125,24 +125,90 @@ func (o *Operator) Kind() OpKind {
 	return o.kind
 }
 
+// Status returns operator status.
+func (o *Operator) Status() OpStatus {
+	return o.status.Status()
+}
+
+// GetReachTimeOf returns the time when operator reaches the given status.
+func (o *Operator) GetReachTimeOf(st OpStatus) time.Time {
+	return o.status.ReachTimeOf(st)
+}
+
+// GetCreateTime gets the create time of operator.
+func (o *Operator) GetCreateTime() time.Time {
+	return o.status.ReachTimeOf(CREATED)
+}
+
 // ElapsedTime returns duration since it was created.
 func (o *Operator) ElapsedTime() time.Duration {
-	return time.Since(o.createTime)
+	return time.Since(o.GetCreateTime())
 }
 
-// RunningTime returns duration since it was promoted.
-func (o *Operator) RunningTime() time.Duration {
-	return time.Since(o.startTime)
+// Start sets the operator to STARTED status, returns whether succeeded.
+func (o *Operator) Start() bool {
+	if o.status.To(STARTED) {
+		atomic.StoreInt64(&o.stepTime, time.Now().UnixNano())
+		return true
+	}
+	return false
 }
 
-// SetStartTime sets the start time for operator.
-func (o *Operator) SetStartTime(t time.Time) {
-	o.startTime = t
+// HasStarted returns whether operator has started.
+func (o *Operator) HasStarted() bool {
+	return !o.GetStartTime().IsZero()
 }
 
-// GetStartTime ges the start time for operator.
+// GetStartTime gets the start time of operator.
 func (o *Operator) GetStartTime() time.Time {
-	return o.startTime
+	return o.status.ReachTimeOf(STARTED)
+}
+
+// RunningTime returns duration since it started.
+func (o *Operator) RunningTime() time.Duration {
+	if o.HasStarted() {
+		return time.Since(o.GetStartTime())
+	}
+	return 0
+}
+
+// IsEnd checks if the operator is at and end status.
+func (o *Operator) IsEnd() bool {
+	return o.status.IsEnd()
+}
+
+// CheckSuccess checks if all steps are finished, and update the status.
+func (o *Operator) CheckSuccess() bool {
+	if atomic.LoadInt32(&o.currentStep) >= int32(len(o.steps)) {
+		return o.status.To(SUCCESS) || o.Status() == SUCCESS
```

Comment:
The second check is used to handle `To(SUCCESS)` failed because of `SUCCESS`. So, I don't think swap is a good idea. Maybe add another `Status() == SUCCESS` before `To(SUCCESS)`? But I don't think efficiency is important here.

Reply:
If so, let keep it.